drivers/soundwire/bus_type.c

```diff
@@ -28,6 +28,14 @@ sdw_get_device_id(struct sdw_slave *slave, struct sdw_driver *drv)
 		id++;
 	}
 
+	/* try and check if this Slave can be handled by a class driver */
```

Comment:
@plbossart i'm probably missing the bigger picture but why do we need the priority if we're first going to match the id for mrfg_id/part_id first and then the class id next?

Reply:
I thought that for each device the kernel would look for all possible drivers, but that's the other way around.
this is counter-intuitive: the match happens by walking through the driver list, and trying to find a device that would work for this driver. Then you try with the next driver, etc.
So if the class driver is first in the list, it will be used by the second part in sdw_get_device_id() for all devices that suppport the class.

Reply:
wow, the linux device-driver model is so complicated. But when does this walk-through happen? 
From what I could tell, the driver probe is triggered by a call to device_register().

Reply:
If when you do a device_register, this will happen, but like I said using a driver iterator:

````c
static int __device_attach(struct device *dev, bool allow_async)
{
	int ret = 0;

	device_lock(dev);
	if (dev->driver) {
		if (device_is_bound(dev)) {
			ret = 1;
			goto out_unlock;
		}
		ret = device_bind_driver(dev);
		if (ret == 0)
			ret = 1;
		else {
			dev->driver = NULL;
			ret = 0;
		}
	} else {
		struct device_attach_data data = {
			.dev = dev,
			.check_async = allow_async,
			.want_async = false,
		};

		if (dev->parent)
			pm_runtime_get_sync(dev->parent);

		ret = bus_for_each_drv(dev->bus, NULL, &data,
					__device_attach_driver);

````

Reply:
@plbossart instead of introducing a priority, can we consider making the match function for soundwire bus a bit smarter ie can we make the class driver relinquish its claim to bind itself to the device because it knows that there's another driver more suitable as well?

Reply:
well, that's the point. it doesn't  know that, it depends when the 'more suitable driver' is loaded or what the other drivers are.

Reply:
Hmm not sure if I'm making sense or maybe it's too complicated. But let's say we have a match with the class_id, can we not iterate through the bus drv's in the match function and see any of the other drivers return a match?

Reply:
isn't that the same as a priority...

Reply:
Yes in theory but it won't need any changes in the device_driver struct. It could be plan B in case the idea of adding a priority doesn't go down well upstream